src/vs/server/node/remoteExtensionHostAgentServer.ts

```diff
@@ -546,21 +548,54 @@ class RemoteExtensionHostAgentServer extends Disposable implements IServerAPI {
 		remoteSocket.pipe(localSocket);
 	}
 
-	private _connectTunnelSocket(host: string, port: number): Promise<net.Socket> {
-		return new Promise<net.Socket>((c, e) => {
-			const socket = net.createConnection(
-				{
-					host: host,
-					port: port
-				}, () => {
-					socket.removeListener('error', e);
-					socket.pause();
-					c(socket);
-				}
-			);
+	private _connectTunnelSocket(host: string, port: number, ct?: CancellationToken): Promise<net.Socket> {
+		const store = new DisposableStore();
+		try {
+			// for localhost, try to use ipv4 and ipv6, resolving whichever works first (#191945)
+			if (host === 'localhost') {
+				const cts = store.add(new CancellationTokenSource(ct));
+				const deferred = new DeferredPromise<net.Socket>();
+				const doResolve = (connection: net.Socket) => {
+					if (deferred.isSettled) {
+						connection.end();
+					} else {
+						deferred.complete(connection);
+						cts.cancel();
+					}
+				};
+				let isSecondFailure = false;
+				const doReject = (err: Error) => {
+					if (!isSecondFailure) {
+						isSecondFailure = true;
+					} else {
+						deferred.error(err);
+					}
+				};
 
-			socket.once('error', e);
-		});
+				this._connectTunnelSocket('127.0.0.1', port, cts.token).then(doResolve, doReject);
+				this._connectTunnelSocket('::1', port, cts.token).then(doResolve, doReject);
+				return deferred.p;
```

Comment:
For apps that listen on all network interfaces, both connections will succeed. I don't feel 100% comfortable attempting to connect to both in parallel so late in endgame, I would prefer a solution where we first try `127.0.0.1`, and if that fails, then try `::1`.